rpm-ostree-bisect

```diff
@@ -0,0 +1,347 @@
+                      'status': 'GOOD',
+                    },
+    }
+"""
+def initialize_commits_info(repo, bad, good):
+    # An ordered dictionary of commit info
+    info = OrderedDict()
+    # The first commit in our list will be the "BAD" commit
+    commitid = bad
+
+    # Iterate over all commits and add them to the array
+    while commitid is not None:
+        _, commit = repo.load_variant(OSTree.ObjectType.COMMIT, commitid)
+        meta = commit.get_child_value(0)
+        version = meta.lookup_value('version', GLib.VariantType.new('s'))
+        info.update({ commitid: { 'version': version.get_string(), 
+                                 'heuristic': 'ASSUMED',
+                                 'status': 'UNKNOWN' }})
+        # Next iteration
+        commitid = OSTree.commit_get_parent(commit)
+
+    # Mark the bad commit bad
+    info[bad]['status'] = 'BAD'
+    info[bad]['heuristic'] = 'GIVEN'
+
+    # Mark the good commit good
+    if good:
+        info[good]['status'] = 'GOOD'
+        info[good]['heuristic'] = 'GIVEN'
+    else:
+        lastcommit = list(info.keys())[-1]
+        info[lastcommit]['status'] = 'GOOD'
+
+    return info
+
+"""
+    Grab all commit history from the remote (just
+    the metadata).
+"""
+def pull_commit_history(deployment, repo):
+
+    # Get repo, remote and refspec from the booted deployment
+    origin = deployment.get_origin()
+    refspec = origin.get_string('origin', 'refspec')
+    # with layered packages it has baserefspec
+    #refspec = origin.get_string('origin', 'baserefspec')
+    _, remote, ref = OSTree.parse_refspec(refspec)
+
+    # Build up options array for pull call
+    #   refs (as): Array of string refs
+    #   flags (i): An instance of OSTree.RepoPullFlags
+    #   depth (i): How far in the history to traverse; default is 0, -1 means infinite
+    # More info on Gvariant types:
+    #   https://lazka.github.io/pgi-docs/GLib-2.0/classes/VariantType.html#GLib.VariantType
+    flags = OSTree.RepoPullFlags(2) # COMMIT_ONLY = 2 Only pull commit metadata
+    depth = -1 # -1 means max depth
+    options = GLib.Variant('a{sv}', {
+        'refs': GLib.Variant('as', [ref]),
+        'flags': GLib.Variant('i', flags),
+        'depth': GLib.Variant('i', depth),
+    })
+
+    # Grab commit metadata history for ref from repo
+    progress = OSTree.AsyncProgress.new()
+    #progress2 = OSTree.AsyncProgress.new_and_connect(OSTree.Repo.pull_default_console_progress_changed(progress, None), None)
+    repo.pull_with_options(remote, options, progress, None) 
+
+def load_data(datafile):
+    with open(datafile, 'r') as f:
+        data = json.load(f, object_pairs_hook=OrderedDict)
+    commits_info = data['commits_info']
+    testscript = data['test_script']
+    return commits_info, testscript
+
+def write_data(datafile, commits_info, testscript):
+    data = { 'commits_info': commits_info,
+             'test_script': testscript }
+    with open(datafile, 'w') as f:
```

Comment:
My "non-transactional file replacement" 🚨 is going off :smile:  - I always do the "create temp then rename" dance.

Reply:
fixed